Update documentation - Optimize index (#217)
microsoft/hyperspace: docs/_docs/03-ug-mutable-dataset.md

## Options of using index when your dataset changes

### Lineage
Hyperspace uses lineage for tracing index entries back to the source data files from which they were generated.
Lineage is required for removing deleted index entries during [index refresh in the incremental mode](#refresh-index---incremental-mode)
or enforcing deletes at the query time when using [Hybrid Scan](#hybrid-scan).
By default, lineage is disabled for indexes, and if required it should be enabled at the time of index creation.
Check the [configuration](https://microsoft.github.io/hyperspace/docs/ug-configuration/) page to see how you can enable lineage for an index.

Once lineage is enabled for an index, Hyperspace adds a new column to the index schema to save the source data file path for each
index entry. This means enabling lineage increases total storage size used for index files, proportional to the

> number of distinct source data files the index is built on. If you know you wont delete any source data file after index

Reviewer: ```suggestion
number of distinct source data files the index is built on. If you know you will not delete any source file after index
```

Author: "source data file" is a term that we are have been using consistently throughout coding, comments and documentation to refer to an index's original dataset files. I would rather keep "source data file" as is.

Reviewer: I made this comment because of https://github.com/microsoft/hyperspace/pull/218#discussion_r509664352

I understand the point - "source data files" is also consistent in our code base and github issues, but
just wanted to be consistent in **user doc** and I think it will look better. WDYT? @pirz

Reviewer: > Let's define two terms clearly and be consistent throughout user documents :)
> 
> - index source file: source files in source relations
> 
> - index data file: index data file that we built

@pirz @imback82 @apoorvedave1 
Ok let's decide "source data files" or "source files" in user doc. I prefer the latter one because it's shorter and clear. But "source data file" would be good as we're using it in the code base and github issue documentation.

creation or you are fine recreating the index using [index refresh in the full mode](#refresh-index---full-mode)

> after deleting some data files, then you dont need to enable lineage for the index. 

Reviewer: ```suggestion
after deleting some files, then you do not need to enable lineage for the index. 
```

After some files are added to or deleted from the original source files, an index was built on,
using `refreshIndex` with the `"incremental"` mode causes the index refresh action recreate any existing index file,
which has some deleted records, to remove those records. Refresh also creates new index files by indexing newly added data files.

> An index needs to have lineage enabled to support deletes during refresh in the incremental mode. Lineage 

Reviewer: Can we create a sub section for lineage (with 6 of #?)? It's pre-requisite for delete refresh/hybrid scan, so I think it's better to make it noticeable.

Author: Sure, but where do you exactly want to put that section? I am thinking it should be with "3 #" and comes before "Options of using index when your dataset changes", as it is used by refresh/hybridScan which come in that section. @sezruby What do you think?

Reviewer: How about "### Pre-requisite" before "### RefreshIndex"?

Author: I added it as "### Lineage"

Reviewer: Thanks! and then now we can remove ... Lineage~ part here and, add link to the section. 

```suggestion
An index needs to have [lineage](#lineage) enabled to support deletes during refresh in the incremental mode. 
```

```

### Optimize Index

> The fastest way to index new data files and merge them into an existing index is by calling `refreshIndex`

Reviewer 2: Maybe change "The fastest way" to "One way" since we will have "quick" mode soon? 😄

Check the [configuration](02-ug-configuration.md) page to see how this threshold can be adjusted.
Quick mode is the default mode for `optimizeIndex`.

Assume you have an index with the name `empIndex`. After adding some data files to the dataset `empIndex` is created on
and refreshing it in the incremental mode, you can optimize `empIndex` in the quick mode as below:

Scala:
```scala
import com.microsoft.hyperspace._

val hs = new Hyperspace(spark)
hs.optimizeIndex("empIndex", "quick")
``` 

Python:

```python
from hyperspace import Hyperspace

hs = Hyperspace(spark)
hs.optimizeIndex("empIndex", "quick")
```

#### Optimize Index - Full Mode
In the `"full"` mode, `optimizeIndex` command considers all existing index files as candidates for merging. Therefore,
Hyperspace does a full scan on current index content and identifies groups of index files which have index records belonging
to the same bucket according to the index configuration. It then replaces each group with a single index file created through
merging all the files in that group together. This mode tries to achieve the best query performance improvement via a

> potentially slow optimize index process.

Reviewer 2: Can you mention the difference between "full" optimize vs. "full" refresh? It is subtle, so it will be worthwhile to call it out.

Author: Added a paragraph on it.

```

### Optimize Index
The fastest way to index newly added source data files and merge them into an existing index is by calling `refreshIndex`
with the `"incremental"` mode. In this mode, each time refresh is called, it creates fresh index files for new data files
and updates index metadata to include them in the index content.

> As these index files accumulate, they could affect query performance when the index is used.

Reviewer: BTW, RefreshDeleteAction will merge them; so append-only incremental refresh ops will make that fragments. I think this could be described in design doc.

Author: Yes, I made a bit of change in the text to emphasize more that index files for appended data cause the issue. I think we should not use `RefreshDeleteAction` term in the doc as that is an internal Action. From a user's perspective, it is only `refreshIndex` API. 
@sezruby Let me know if you still think more context is required here.

Reviewer: Yes this might be enough for user document.

Once the index is leveraged for a query,
the large number of these files could increase overall query time as more index files need to be accessed and potentially
read to compute the query results.

> Hyperspace provides the `optimizeIndex` command to alleviate above issue by changing index files layout for an

Reviewer: I think we need to show the status of index data files (e.g. number of files) and suggest to run optimizeIndex so that a user can easily determine and trigger it if needed. Could you create an issue for this?

Author: Sure, are suggesting to add a new API to Hyperspace which can be called directly by user and shows some stats on the index, including how "fragmented" it is?

Reviewer: Yes we could show 
- general index stats
  - name, index data bytes,
  - all versions for index data files of (latest version?)
     - can be multiple in case optimizeIndex or incremental refresh)
- source data context - (we could merge with isValid (#22))
  - number of current files / appended files / deleted files 
  - show 3~5 full paths of appended or deleted fils 
- optimize context
  - how many optimize-ndeeded files and their size
  - recommend optimizeIndex or not?
- root paths of index source files 

I think we could do this with 
hs.getIndexStatus(IndexName) or just hs.optimizeIndex(IndexName, mode, dryRun = true) only for "optimize context"?
WDYT? @pirz

Author: @sezruby - Overall looks fine to me. I just think:
1) We should keep it as a separate command (like what you suggest: `getIndexStatus(ixName)` - or probably rename it with sth else) and we should not mix it with Optimize.
2) "recommend optimizeIndex or not?": I am not sure if it is a good idea to make a such a recommendation in the output of such an API. Normally, a "getStatus" type of API should be metadata read/summarization only. Making such a recommendation needs some (deeper) index content analysis. We need to check how much of overlap we have between new index file's buckets and the older ones. This does not necessarily need scanning index records, but can be done by extracting bucket-ids from file names and grouping them on that to see how many buckets with ">1 files" are there. But in general, "getStatus()" APIs and "recommend()" APIs are normally separated. 
Adding @rapoth - @apoorvedave1 - @imback82 to make suggestions

index which has many index files, due to incremental refresh call(s). This is achieved by merging index files together,
if possible, and replacing them with fewer larger files that capture exact same index records. This process is similar to
compaction in append-only log structured merge index structures.

You should note that the `optimizeIndex` command is a best effort to modify index files layout and its
final outcome depends on how index records are stored in existing index files.
Two or more index files can be merged with each other, if and only if they all have index records which belong to the
same bucket (according to index configuration).
When running `optimizeIndex` command, Hyperspace tries to find such groups of index files and merge them together.

> However, if most of the index files have records from different buckets, then final physical layout of index files

Reviewer: > if most of the index files have records from different buckets, 

Could you rephrase this part?

Author: I made a bit of change here. Please let me know if that addressed your comment.

The size threshold for an index file to be eligible for merging during quick optimization can be changed.
Check the [configuration](02-ug-configuration.md) page to see how this threshold can be adjusted.
Quick mode is the default mode for `optimizeIndex`.

Assume you have an index with the name `empIndex`. After adding some data files to the dataset `empIndex` is created on
and refreshing it in the incremental mode, you can optimize `empIndex` in the quick mode as below:

Scala:
```scala
import com.microsoft.hyperspace._

val hs = new Hyperspace(spark)
hs.optimizeIndex("empIndex", "quick")
``` 

Python:

```python
from hyperspace import Hyperspace

hs = Hyperspace(spark)
hs.optimizeIndex("empIndex", "quick")
```

#### Optimize Index - Full Mode
In the `"full"` mode, `optimizeIndex` command considers all existing index files as candidates for merging. Therefore,
Hyperspace does a full scan on current index content and identifies groups of index files which have index records belonging
to the same bucket according to the index configuration. It then replaces each group with a single index file created through

> merging all the files in that group together. This mode tries  to achieve the best query performance improvement via a

Reviewer: ```suggestion
merging all the files in that group together. This mode tries to achieve the best query performance improvement via a
```

Two or more index files can be merged with each other, if and only if they all have index records which belong to the
same bucket (according to index configuration).
When running `optimizeIndex` command, Hyperspace tries to find such groups of index files and merge them together.

> However, if for each given bucket, all index records are already stored in a single index file (i.e. we dont have several index

Reviewer: ` (i.e. it does not have several` ?

Reviewer: How about `already stored in a single index file` => `already stored in a single index file for each bucket`??

Author: I revised that part again. Plz take a look at it.

> files for a bucket), then final physical layout of index files after running optimize would be similar to the original

Reviewer: I think it should be the same if all index records are already stored in a single index file for each bucket, we shouldn't change it. (Though current optimizeIndex rewrites the data and change it, but we don't need to describe it here.) and "final physical layout" term could be confusing. 

How about just removing "However ~~~ original layout" lines?

Author: I revised that part again. Plz take a look at it.
